### PR TITLE
FCM support

### LINKED ISSUE #48: Push Notification
Extend the drop server that he can send Drop Messages via push notifications to a client.

### PR DESCRIPTION
**Pull request by enkore**
Fixes #48

### CODE REVIEW

**Comment by enkore** [comment]
This branch has conflicts that must be resolved *sigh*

**Comment by enkore** [comment]
FCM works API-wise. It uses topic messages, so message routing is handled by FCM and no explicit routing and subscription-keeping is required on our side. The app just subscribes to the drop-ids it wants and that should be it.

The FCM API looks kinda slow from where I'm standing. Observe this histogram:

```
notify_fcm_duration_bucket{le="0.25"} 0.0
notify_fcm_duration_bucket{le="0.5"} 3.0
notify_fcm_duration_bucket{le="0.75"} 4.0
notify_fcm_duration_bucket{le="1.0"} 5.0
```

The pure API calls block the worker process for ~0.5 s.

WebSocket support via dwr didn't work reliably at any point. Either it sporadically lost messages, deadlocked the worker or just didn't work at all. At this point I'm almost considering just moving the entire server to an asynchronous base supporting WS out of the box. That would also have the advantage that theFCM calls would only block an asyncore, *not the entire process*.

File: drop_service/monitoring.py [hunk 1]
```diff
@@ -1,7 +1,27 @@
...
 
 DROP_RECEIVED = Counter('drop_messages_received',
                         'Amount of received drop messages')
+
+
+FCM_CALLS = Counter('notify_fcm_calls',
+                    'Number of notify_topic_subscribers FCM API calls', ['exception'])
+
+
+FCM_DURATION = Histogram('notify_fcm_duration',
```
Comment (audax) [inline drop_service/monitoring.py]: Use a Summary() instead of a counter and a histogram.

File: drop_service/notify.py [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+
+class FCM:
+    """
+    Publish drops through FCM topic messages. The topic is the drop ID.
+    """
+    SERVICE = 'fcm'
+    _logger = logger.getChild('fcm')
+
+    def __init__(self):
+        self._push = FCMNotification(api_key=settings.FCM_API_KEY, proxy_dict=settings.FCM_PROXY)
```
Comment (audax) [inline drop_service/notify.py]: Inject FCMNotification in the constructor (maybe with a default value that acts as a fallback) to test this class

**Comment by enkore** [comment]
```
django_http_requests_latency_including_middlewares_seconds_bucket{le="0.005"} 4.0
django_http_requests_latency_including_middlewares_seconds_bucket{le="0.01"} 862.0
django_http_requests_latency_including_middlewares_seconds_bucket{le="0.025"} 1488.0
django_http_requests_latency_including_middlewares_seconds_bucket{le="0.05"} 1581.0
django_http_requests_latency_including_middlewares_seconds_bucket{le="0.075"} 1588.0
django_http_requests_latency_including_middlewares_seconds_bucket{le="0.1"} 1590.0

process_cpu_seconds_total 23.91
```

I can live with that for now.

**Comment by enkore** [comment]
> btw. Summary doesn't seem to generate the quartiles/buckets here:

Prometheus docs:

> First of all, check the library support for histograms and summaries. Full support for both currently only exists in the Go client library. Many libraries support only one of the two types, or they support summaries only in a limited fashion (lacking quantile calculation).

I guess we go back to Counter/Histogram combination or maybe just implement that. It's useful.

**Pull request merged**